engine/gamesys/src/gamesys/scripts/script_resource.cpp

```diff
@@ -842,6 +878,707 @@ static int SetTexture(lua_State* L)
+ *             }
+ *         }
+ *     }
+ *     resource.set_atlas("/main/test.a.texturesetc", params)
+ * end
+ * ```
+ */
+
+static int SetAtlas(lua_State* L)
+{
+    DM_LUA_STACK_CHECK(L, 0);
+
+    dmhash_t path_hash = dmScript::CheckHashOrString(L, 1);
+    void* atlas_res    = CheckResource(L, g_ResourceModule.m_Factory, path_hash, "texturesetc");
+    assert(atlas_res);
+
+    dmGameSystemDDF::TextureSet texture_set_ddf = {};
+    uint32_t frame_index_count                  = 0;
+    uint32_t num_geometries                     = 0;
+    uint32_t num_animations                     = 0;
+    float tex_width                             = 0;
+    float tex_height                            = 0;
+
+    luaL_checktype(L, 2, LUA_TTABLE);
+    lua_pushvalue(L, 2);
+
+    // Set texture resource path
+    lua_getfield(L, -1, "texture");
+    {
+        const char* texture_path         = luaL_checkstring(L, -1);
+        dmhash_t tex_canonical_path_hash = GetCanonicalPathHash(texture_path);
+        void* texture_res                = CheckResource(L, g_ResourceModule.m_Factory, tex_canonical_path_hash, "texturec");
+        assert(texture_res);
+        texture_set_ddf.m_Texture        = texture_path;
+        dmGraphics::HTexture texture     = (dmGraphics::HTexture) texture_res;
+        tex_width                        = dmGraphics::GetTextureWidth(texture);
+        tex_height                       = dmGraphics::GetTextureHeight(texture);
+    }
+    lua_pop(L, 1); // "texture"
+
+    // Note: We do separate passes over the lua state to validate the data in the args table,
+    //       this is because we need to allocate dynamic memory and can't use luaL_check** functions
+    //       since they longjmp away so we can't release the memory..
+    ValidateSetAtlasGeometries(L, &num_geometries);
+    ValidateSetAtlasAnimations(L, &num_animations);
+
+    if (num_animations > 0 && num_geometries == 0)
+    {
+        lua_pop(L, 1);
+        return DM_LUA_ERROR("Setting atlas with animations requires geometry");
+    }
+
+    texture_set_ddf.m_Geometries.m_Data  = new dmGameSystemDDF::SpriteGeometry[num_geometries];
+    texture_set_ddf.m_Geometries.m_Count = num_geometries;
+    memset(texture_set_ddf.m_Geometries.m_Data, 0, sizeof(dmGameSystemDDF::SpriteGeometry) * num_geometries);
+
+    texture_set_ddf.m_Animations.m_Data  = new dmGameSystemDDF::TextureSetAnimation[num_animations];
+    texture_set_ddf.m_Animations.m_Count = num_animations;
+    memset(texture_set_ddf.m_Animations.m_Data, 0, sizeof(dmGameSystemDDF::TextureSetAnimation) * num_animations);
+
+    if (num_geometries > 0)
+    {
+        float inv_tex_width  = 1.0f / tex_width;
+        float inv_tex_height = 1.0f / tex_height;
+
+        lua_getfield(L, -1, "geometries");
+        for (int i = 0; i < num_geometries; ++i)
+        {
+            lua_pushnumber(L, i+1);
+            lua_gettable(L, -2);
+            assert(lua_istable(L, -1));
```

Comment:
Is this a programmer error or a user error?
If they're playing with the new api, and suddenly the app dies, it feels like a bad alternative.

At the same time, I'm not advocating for sprinkling our code with defensive checks, but I think making the app close is perhaps a bit much.

Reply:
It's more of a programmer error used for developing the feature as I felt it's really easy to mess up the lua stack when doing nested iteration, I think we can remove it tbh